enhancements/installer/feature-sets.md

```diff
@@ -0,0 +1,224 @@
+
+### Non-Goals
+
+None
+
+## Proposal
+
+### Workflow Description
+
+Example workflow demonstrating feature set validation and use:
+
+**Developer** contributes code to the openshift-install codebase.
+
+**Admin** is the user performing an install.
+
+1. Developer adds feature `foo` to openshift-install. Feature `foo`
+includes field `bar` in the install config; feature `foo` is part of the Tech Preview Feature Set.
+2. Admin specifies `bar: baz` in install-config.yaml (but `featureSet` is empty).
+3. `openshift-install` returns an error: `the TechPreviewNoUpgrade feature set must be enabled to use this field`
+4. Admin adds `featureSet: TechPreviewNoUpgrade` to install-config.yaml
+5. Installer generates `FeatureGate` manifest
+6. Installer emits warning message indicating the enabled feature set when provisioning infrastructure
+7. Cluster installs successfully with TechPreview cluster 
+
+
+### API Extensions
+
+[OpenShift API Feature Sets](https://github.com/openshift/api/blob/master/config/v1/types_feature.go)
+would become part of the install config API:
+
+```go
+
+// FeatureSet enables features that are not part of the default feature set.
+	// +optional
+	FeatureSet configv1.FeatureSet `json:"featureSet,omitempty"`
+```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Install Config
+
+The basic implementation through the install config is relatively straight forward. Gated features are protected
+by validation:
+
+```go
+
+	if c.FeatureSet == nil || *c.FeatureSet != types.TechPreviewNoUpgrade {
+		errMsg := "the TechPreviewNoUpgrade feature set must be enabled to use this field"
+
+    if installConfig.Foo != nil && len(installConfig.Foo.Bar) > 0 {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("foo", "bar"), errMsg))
+		}
+	}
+```
+
+Package `types` defines the install config and represents the API offered by the Installer. Most notably this
+API is consumed/used by hive. For this reason, it is important that feature sets be part of the install config.
+
+#### CLI Flag (future work)
+
+On the other hand, the install config is an asset and gated features could potentially be exercised
+before the install config is validated (or perhaps the feature is completely independent from the install config). For this reason, it would
+be a good idea to provide a flag:
+
+```shell
+$ openshift-install create cluster --feature-set TechPreviewNoUpgrade
+```
+
+This flag, if provided, would also populate the value in the install config; and conflicts between the flag and install config
+would throw an error.
+
+While there are clear use cases for a flag, there are no immediate use cases and the implementation of the Installer (with
+its separation between command-line flags and assets) does not lead to a clear-cut design. Implementing this functionality
+is certainly possible, but I propose we defer implementation for this until it is determined to be necessary.
```

Comment:
No changes to this necessary, but if we pursue this I'd assume we want this flag to go back through the asset pipeline ensuring that TechPreviewNoUpgrade featurset is defined in cluster as well.